torch/csrc/api/include/torch/data/ctf/ctf_parser.h

```diff
@@ -0,0 +1,931 @@
+#pragma once
+
+#include <torch/data/ctf/reader.h>
+#include <torch/data/ctf/utils.h>
+
+#include <stdint.h>
+#include <algorithm>
+#include <cassert>
+#include <climits>
+#include <iostream>
+#ifdef CTF_DEBUG
+#include <map>
+#endif
+#include <memory>
+#include <ostream>
+#include <unordered_map>
+#include <vector>
+
+namespace torch {
+namespace data {
+namespace ctf {
+
+/*
+ * CTF general format
+ * [Sequence_Id](Sample or Comment)+
+ *   where
+ *          sequence_Id=(empty|[0-9]+)
+ *          Sample=|Input_Name (Value )*
+ *          Comment=|# some content
+ * Example:
+ * 100 |a 1 2 3 |b 100 200
+ * 100 |a 4 5 6 |b 101 201
+ * 100 |b 102983 14532 |a 7 8 9
+ * 100 |a 7 8 9
+ * 200 |b 300 400 |a 10 20 30
+ * 333 |b 500 100
+ * 333 |b 600 -900
+ * 400 |a 1 2 3 |b 100 200
+ * |a 4 5 6 |b 101 201
+ * |a 4 5 6 |b 101 201
+ * 500 |a 1 2 3 |b 100 200
+ */
+
+///
+/// Beginning of type definitions
+///
+
+///
+/// Enumeration type denoting data type of symbolic data entities or actual
+/// data.
+///
+enum class CTFDataType : unsigned int {
+  Unknown = 0,
+  Float = 1,
+  Double = 2,
+  UChar = 3, // So far only used internally in deserializers.
+  Float16 = 4,
+  Int8 = 5,
+  Int16 = 6,
+  Int32 = 7,
+};
+
+///
+/// Enumeration type denoting the format of storage
+///
+enum class CTFDataStorage { Dense, Sparse };
+enum class CTFInputStreamType { Feature, Label };
+
+///
+/// Input Stream information
+///
+struct CTFInputStreamInformation {
+  // Self-assigned Unique ID of the input stream (do not assign it!)
+  // TODO: ugly, fix this!
+  size_t __id__;
+  // Unique name of the input stream
+  std::string name;
+  // Unique alias of the input
+  // Useful when the name is long
+  std::string alias;
+  // expected number of elements in a sample
+  // TODO: Only useful if number of samples is known
+  size_t dimension;
+  // Input streams belong to either Feature or Label
+  CTFInputStreamType type;
+  // Data storage of the stream
+  CTFDataStorage storage;
+
+  CTFInputStreamInformation(
+      std::string name,
+      std::string alias,
+      size_t dimension,
+      CTFInputStreamType type,
+      CTFDataStorage storage)
+      : name(std::move(name)),
+        alias(std::move(alias)),
+        dimension(dimension),
+        type(type),
+        storage(storage){};
+
+  // Used for unit tests
+  CTFInputStreamInformation(
+      size_t id,
+      std::string name,
+      std::string alias,
+      size_t dimension,
+      CTFInputStreamType type,
+      CTFDataStorage storage)
+      : __id__(id),
+        name(std::move(name)),
+        alias(std::move(alias)),
+        dimension(dimension),
+        type(type),
+        storage(storage){};
+};
+inline bool operator==(
+    const CTFInputStreamInformation& lhs,
+    const CTFInputStreamInformation& rhs) {
+  return (
+      lhs.__id__ == rhs.__id__ && lhs.name == rhs.name &&
```

Comment:
We can simplify the logic here by returning std::memcmp(&lhs, &rhs, sizeof(lhs))==0